keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+provided together with a DRA driver. As soon as such a custom CR is created, a
+CRD controller of the DRA driver must convert it into a in-tree claim parameter
+object that uses the "parameter types" defined by the semantic models.
+
+The scheduler receives both the node resource information and the claim
+parameters. It uses the in-tree claim parameters to identify and allocate
+suitable resources. It captures all information that the DRA driver kubelet
+plugin will need in the "allocation types" of the semantic models and stores
+that in the claim status. Then it schedules the pod.
+
+The DRA driver kubelet plugin receives that information when the pod is about to
+start. It checks that the hardware is still available and healthy, then
+configures it for the pod as specified in the allocation types.
+
+### Publishing node resources
+
+A semantic model defines an attributes type. This is a Go type that gets
+embedded inside the resource.k8s.io API, not a separate API kind. The kubelet
+queries the registered DRA driver kubelet plugins for resource information
+through a new gRPC call. A streaming gRPC API allows drivers to update the
+capacity information.
+
+For example, a simple counter could be described in a `counter` Go package as:
+```
+type Attributes struct {
+    // Some meta information just for logging and debugging. Not used for counting.
+	metav1.ObjectMeta
+
+	Count int64
+}
+```
+
+The kubelet then publishes this information in a new NodeResources object on
+behalf of the driver. This implies that kubelet, DRA driver, and apiserver must
+all support the same semantic models and the same versions of them. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+### Using semantic parameters as claim parameters
+
+The semantic model also defines a type for claim parameters. For the example
+above, that could be:
+```
+type Parameters struct {
+	Required int64
+}
+```
+
+The `Parameters.Required` value gets compared against `Attributes.Count`.
+
+Suppose a vendor has a CRD which allows requesting a certain number of `frobnicators`:
+
+```
+kind: ClaimParameters
+apiVersion: dra.frobincatorInc.example.com/v1alpha1
+metadata:
+  name: my-claim-parameters
+  uid: abcd
+  resourceVersion: 123
+spec:
+  requiredFrobincators: 100
+```
+
+The vendor CRD controller needs to sync this CRs into one in-tree object:
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+metadata:
+  name: frobnicator-claim-parameters
+  namespace: default
+generatedFrom:
+  kind: ClaimParameters
+  apiGroup: dra.frobincatorInc.example.com
+  name: my-claim-parameters
+  resourceVersion: 123
+  uid: abcd
+parameters:
+  - counter:
+      required: 100
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list
```

Comment:
This is not true anymore, right? The DRA controller now will publish a `ClaimParameters` object; scheduler only needs to be able to read that.

Reply:
It's still true. I added one section about the why below: the kubelet plugin needs the configuration parameters. Those are only in the opaque vendor CRD. A snapshot of that must be taken during allocation and passed through. This applies to claim and class parameter objects.

This solves two problems (outlined already in the core DRA):
- An allocated claim remains usable even if the user then (intentionally or accidentally) removes the claim parameter object. This is particularly important when a pod stops running an kubelet needs to clean up.
- kubelet and kubelet plugins don't need permission to read the vendor CRDs (more secure this way).

Reply:
So, talking with @alculquicondor we want to see if we can avoid scheduler reading all this stuff.

If the DRA controller publishes the vendor-neutral `ClaimParameters` object, can we stuff the CR in there as a RawExtension? So, the version used to create the `ClaimParameters` is right there along side it. No need to guess about whether it's been modified. When the scheduler makes the allocation, it can copy the `ClaimParameters` into allocation result. Wouldn't that solve it without immutability, and without reading arbitrary resources in the scheduler?

Reply:
Vendor neutral ClaimParameters sounds ok to me.
Another alternative is this https://github.com/kubernetes/enhancements/pull/4384#discussion_r1480352958

Reply:
> If the DRA controller publishes the vendor-neutral ClaimParameters object, can we stuff the CR in there as a RawExtension?

Yes. I came to the same conclusion after missing this comment.

We need to do the same for resource class parameters, but that's the same concept.

Reply:
Awesome. That cleans things up a lot, I think.

Reply:
Note that we loose the ability to check the generated parameters for "up-to-date-ness" because the scheduler is never reading the original object and thus doesn't know the UID and Generation of it. But that's okay: if users are updating claim parameters while allocation is going on, there's always a race somewhere. This cannot be prevented by the system.

Reply:
RawExtension for vendor configuration parameter in ResourceClaimParameters was added. Same for resource class parameters.